F/169 add telemetry and consent notice config

rweber-esri opened the pull request:
…new site.data.values.telemetry object. Additionally, prevent the schemaVersion from being incremented when only updating published status when saving a draft.

[169](https://devtopia.esri.com/dc/hub/issues/169)

1) Adds new `site.data.values.telemetry` config object
2) Move existing `gacode` under `site.data.values.telemetry.customAnalytics.ga`
3) Fixes bug where saving a site draft can cause loss of migration changes

tomwayson commented on packages/sites/src/index.ts:
```diff
@@ -58,3 +58,4 @@ export * from "./get-data-for-site-item";
 export * from "./get-site-by-id";
 export * from "./is-site";
 export * from "./get-members";
+export * from "./_ensure-telemetry";
```
should this be exported here? It appears this is a "private" fn that we only export for ease of testing.

rweber-esri replied:
I only added it b/c I noticed other private method & migration module exports, doesn't need exported, so happy to remove

rweber-esri replied:
Scratch that, deleting the export statement breaks tests that try to mock out this migration

tomwayson commented on packages/sites/src/drafts/save-published-status.ts:
```diff
@@ -20,7 +21,17 @@ export function savePublishedStatus(
+    // when saving a draft site, we need to prevent the schemaVersion
+    // from being updated. if not, and the user refreshes the
+    // browser after saving the draft without first publishing,
+    // migration changes can be lost
```
I don't quite follow how the changes would be lost, but I trust you.

dbouwman replied:
I'm not following this either... @rweber-esri can you explain what at risk of getting lost?

rweber-esri replied:
This resolves an issue I noticed that can be reproduced by the following steps:
1. Site is at schemaVersion 1.3 in AGO
2. User navigates to /edit, for which schema 1.4 is applied, but not persisted to AGO.
3. User makes local changes & saves draft.
4. When the above method is invoked, we make `typeKeywords` the only thing in the `allowList`, but the `updateSite` method automatically adds `schemaVersion` to the `allowList`, which results in `schemaVersion` being bumped to 1.4 & the typeKeyword representing we have an unpublished draft is added to the item, but none of the migration changes are committed as a result.
5. User refreshes their browser, now the item in AGO has schemaVersion 1.4, so the migration isn't applied.
6. The user saves another draft
7. Migrations changes are lost to the wind

rweber-esri replied:
Updated the comment to be more accurate wrt the STR in my previous comment

tomwayson commented on packages/sites/src/update-site.ts:
```diff
@@ -21,11 +21,13 @@ import { updateItem, IUpdateItemResponse } from "@esri/arcgis-rest-portal";
   patchList: string[],
-  hubRequestOptions: IHubRequestOptions
+  hubRequestOptions: IHubRequestOptions,
+  updateVersions: boolean = true
```
is there another way we can achieve the same w/o adding a trailing boolean argument? We try to have request options as the final argument. Maybe have `IUpdateSiteOptions extends IHubRequestOptions` that has `updateVersions` in it (also should ultimately have `patchList` in it too)?

dbouwman replied:
+1 for `IUpdateSiteOptions` vs trailing boolean

rweber-esri replied:
I was going for a non-breaking change to avoid a major bump, but agree, the proposed is better. Looks reasonably easy to account for in Hub & Solutions.js.

tomwayson reviewed (commented):
Nice work @rweber-esri. I just wonder if there's another way we can deal w/ `updateVersions`.

rweber-esri commented on packages/sites/src/update-site.ts:
```diff
@@ -15,73 +16,73 @@ import { updateItem, IUpdateItemResponse } from "@esri/arcgis-rest-portal";
- * @param {IHubRequestOptions} hubRequestOptions
+ * @param {IUpdateSiteOptions} updateSiteOptions
  */
 export function updateSite(
```
I don't disagree, but it looks like [update-page](https://github.com/Esri/hub.js/blob/master/packages/sites/src/pages/update-page.ts#L32) already differs from `update-page` in that it can receive 2 or 3 args and there's some variable reassignment to account for this.

dbouwman replied:
In an ideal world, we'd keep these signatures pretty similar... but they are different item types, so I don't think we should get too wound up about it

rweber-esri replied:
Thanks for your input @dbouwman. Will hold off on making this change for now then.

tomwayson replied:
FYI - though, this was already on the [list of things to change at the next breaking version](https://github.com/Esri/hub.js/issues/305): 

https://github.com/Esri/hub.js/blob/master/packages/sites/src/pages/update-page.ts#L32-L35

rweber-esri replied:
Pushed another commit to add `IUpdatePageOptions` and refactor `updatePage` to use per side-convo w/ @tomwayson

rweber-esri commented:
I believe I've addressed all PR feedback. Hoping to get this merged & published today, scheduled to go out w/ tomorrow's release & need to huck a Solutions.js PR to bump their dependency on this package beforehand. 🤞 Any additional feedback? Not sure if @tomwayson would like to take one last peak?

tomwayson reviewed (approved):
Thanks for making those changes @rweber-esri
